api/enki.py

```diff
@@ -249,17 +294,136 @@ def checkout(self, patron, pin, licensepool, internal_format):
+        response = self.loan_request(patron.authorization_identifier, pin, book_id)
+        if response.status_code != 200:
+            raise CannotFulfill(response.status_code)
+        result = json.loads(response.content)['result']
+        if not result['success']:
+            message = result['message']
+            if "That record is already checked out to you" in message:
+                # this might be okay
+                pass
+            elif "There are no available copies" in message:
+                # TODO: raise no copies error
+                self.log.error("There are no copies of this book available.")
+                raise CirculationException()
+            elif "Login unsuccessful" in message:
+                # TODO: raise invalid barcode/password error
+                self.log.error("User validation against Enki server was unsuccessful.")
+                raise CirculationException()
+        media_type = self.get_enki_media_type(book_id)
+        url, item_type, expires = self.parse_fulfill_result(result)
+
+        if not media_type and item_type == 'epub':
+            media_type = self.epub
+
+        return FulfillmentInfo(
+            licensepool.collection,
+            licensepool.data_source.name,
+            licensepool.identifier.type,
+            licensepool.identifier.identifier,
+            content_link=url,
+            content_type=media_type,
+            content=None,
+            content_expires=expires
+        )
+
+    def get_enki_media_type(self, book_id):
+        url = str(self.base_url) + str(self.item_endpoint)
+        args = dict()
+        args['method'] = 'getItem'
+        args['recordid'] = book_id
+        args['lib'] = self.library_id
         response = self.request(url, method='get', params=args)
+        if response.status_code != 200:
+            return None
+        media_type = json.loads(response.content)['result']['availability']['accessType']
+        if media_type == 'acs':
+            return self.adobe_drm
+        elif media_type == 'free':
+            return self.no_drm
+        else:
+            return None
+
+    def parse_fulfill_result(self, result):
+        links = result['checkedOutItems'][0]['links'][0]
+        url = links['url']
+        item_type = links['item_type']
+        due_date = result['checkedOutItems'][0]['duedate']
+        expires = self.epoch_to_struct(due_date)
+        return (url, item_type, expires)
+
+    def patron_activity(self, patron, pin):
+        response = self.patron_request(patron.authorization_identifier, pin)
+        if response.status_code != 200:
+            raise PatronNotFoundOnRemote(response.status_code)
+        result = json.loads(response.content)['result']
+        if not result['success']:
+            message = result['message']
+            if "That record is already checked out to you" in message:
+                # this might be okay
+                pass
```

Comment:
Why would either of these messages happen? You're not trying to create a loan or get a book. If this is copy-and-paste I think you can remove it.